Engine Merge obeys row policy

ilejn opened the pull request:
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Engine `Merge` filters the records according to the row policies of the underlying tables.

ilejn committed:
merge_row_policy: initial

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #50209 from ilejn/ClickHouse:merge_row_policy -->
This is an automated comment for commit 81729cbed6ca6939fadbea8d157316a62c51cba1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50209/81729cbed6ca6939fadbea8d157316a62c51cba1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

ilejn committed:
merge_row_policy: tiny cleanup

ilejn committed:
merge_row_policy: with QueryProcessingStage::Complete

ilejn committed:
merge_row_policy - extra debug

ilejn committed:
merge_row_policy: original behavior restored + extra debug

ilejn committed:
merge_row_policy: FilterTransform

ilejn committed:
merge_row_policy: row_policy_storage_merge test

ilejn committed:
merge_row_policy: style fixes and back to storage->read

ilejn committed:
merge_row_policy: addFilter() instead of FilterTransform

ilejn committed:
merge_row_policy: add_filter() actually works

ilejn committed:
merge_row_policy: addFilter(() together with FilterTransform

ilejn commented:
I believe that test failures in 00002_log_and_exception_messages_formatting are irrelevant, although AST fuzzer (ubsan) issue IS relevant.
Fuzzer generated a query
`
SELECT rand(NULL) % NULL FROM merge(currentDatabase(), '02763_merge_merge_1') PREWHERE (rand(1.1920928955078125e-7) % -0.) % rand(NULL) ORDER BY rand(2) % -1 DESC NULLS FIRST, x DESC NULLS FIRST SETTINGS optimize_move_to_prewhere = 0`
which is wrong and produces 
`Code: 59. DB::Exception: Received from localhost:9000. DB::Exception: Illegal type Float64 of column for filter. Must be UInt8 or Nullable(UInt8) or Const variants of them.: While executing MergeTreeInOrder. (ILLEGAL_TYPE_OF_COLUMN_FOR_FILTER)`
on my laptop.
Looking into why it crashes in CI environment.

ilejn committed:
merge_row_policy: cleanup, remove some debug output

ilejn committed:
merge_row_policy: further cleanup

ilejn committed:
merge_row_policy: more tests

Enmk commented on src/Storages/StorageMerge.cpp:
```diff
@@ -666,6 +669,44 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+                        required_columns.begin(), required_columns.end(),
+                        std::inserter(filter_columns, filter_columns.begin()));
+
+                    source_step_with_filter->addFilter(filter_dag_ptr, filter_columns.front());
```
Why only with columns from RBAC that are not required columns? Would filter work properly if there are some required columns in expression?

Enmk replied:
Why only 1 filter, what if there are many columns?

ilejn replied:
The 'extra' column here is a complex condition that may contain logical operations, something like
'field1 > x and field2 > y'.
That's why only one filter is needed.
I'll try to simulate collision between the column and what is listed in required column, this is an interesting point.

Enmk replied:
I believe that it might be beneficial to put that info into comment for this piece of code

Enmk commented on src/Storages/StorageMerge.cpp:
```diff
@@ -666,6 +669,44 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+                    Names filter_columns;
+
+                    std::set_difference(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end(),
+                        required_columns.begin(), required_columns.end(),
```
```suggestion
                        required_columns_sorted.begin(), required_columns_sorted.end(),
```

Enmk commented on src/Storages/StorageMerge.cpp:
```diff
@@ -1001,6 +1047,43 @@ void ReadFromMerge::convertingSourceStream(
+
+
+        std::set_difference(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end(),
+            required_columns.begin(), required_columns.end(),
```
Same here, `required_columns_sorted` ?

Enmk commented on src/Storages/StorageMerge.cpp:
```diff
@@ -1001,6 +1047,43 @@ void ReadFromMerge::convertingSourceStream(
+
+    auto row_policy_filter = local_context->getRowPolicyFilter(database_name, table_name, RowPolicyFilterType::SELECT_FILTER);
+
+    if (row_policy_filter)
+    {
+        ASTPtr expr = row_policy_filter->expression;
+
+        auto syntax_result = TreeRewriter(local_context).analyze(expr, pipe_columns);
+        auto expression_analyzer = ExpressionAnalyzer{row_policy_filter->expression, syntax_result, local_context};
+
+        auto actions_dag = expression_analyzer.getActionsDAG(true, false);
+        auto filter_actions = std::make_shared<ExpressionActions>(actions_dag, ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+        auto required_columns = filter_actions->getRequiredColumns();
+
+        LOG_TRACE(&Poco::Logger::get("ReadFromMerge::convertinfSourceStream"), "filter_actions_dag: {},<> {}, <> {}",
+            filter_actions->getActionsDAG().dumpNames(), filter_actions->getActionsDAG().dumpDAG(), filter_actions->getSampleBlock().dumpStructure());
+
+
+        auto fa_actions_columns_sorted = filter_actions->getSampleBlock().getNames();
+        std::sort(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end());
+
+        Names required_columns_sorted = required_columns;
+        std::sort(required_columns_sorted.begin(), required_columns_sorted.end());
+
+        Names filter_columns;
+
+
+        std::set_difference(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end(),
+            required_columns.begin(), required_columns.end(),
+            std::inserter(filter_columns, filter_columns.begin()));
```
This looks almost identical to the code block you've added above, maybe factor it out into a function?

Enmk commented on tests/queries/0_stateless/02763_row_policy_storage_merge.sql.j2:
Maybe also add RBAC based on 2 or more columns?

Enmk reviewed (changes requested):
Please address the comments

ilejn committed:
merge_row_policy: namesDifference, try to handle nonselected columns

ilejn committed:
merge_row_policy: works again with adding missed columns

ilejn committed:
merge_row_policy: refactored with ReadFromMerge::RowPolicyData

ilejn committed:
merge_row_policy: stylecheck

ilejn committed:
merge_row_policy: stylecheck again

ilejn committed:
merge_row_policy: cleanup, comments, new tests

Enmk commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +850,113 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
 }
 
+void ReadFromMerge::RowPolicyData::init(RowPolicyFilterPtr row_policy_filter_ptr_,
+    const std::shared_ptr<DB::IStorage> storage,
```
Please remove `const` here, it breaks the build.

Enmk commented on src/Storages/StorageMerge.cpp:
```diff
@@ -961,8 +1129,14 @@ void ReadFromMerge::convertingSourceStream(
     ContextPtr local_context,
     QueryPipelineBuilder & builder,
-    const QueryProcessingStage::Enum & processed_stage)
+    const QueryProcessingStage::Enum & processed_stage,
```
Please remove `const` here too

ilejn committed:
merge_row_policy: cleanup, verbose short messages in 00002_log ...

ilejn committed:
merge_row_policy: more verbosity for 00002_log_and_exception_...

ilejn committed:
merge_row_policy: revert debug, 00002_log... test is fixed in master

ilejn commented:
Integration test failures are not related.

Please, review proposed change.

ilejn commented:
Hello @vitlibar ,
could you please review this PR and/or attract someone's attention to it?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,113 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+        actions_dag = expression_analyzer.getActionsDAG(true, false);
+        filter_actions = std::make_shared<ExpressionActions>(actions_dag,
+            ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+        filter_column_name = namesDifference(filter_actions->getSampleBlock().getNames(), filter_actions->getRequiredColumns());
```
What is `filter_column_name`?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,113 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+        auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
+        auto expression_analyzer = ExpressionAnalyzer{expr, syntax_result, local_context};
+
+        actions_dag = expression_analyzer.getActionsDAG(true, false);
```
Why `(true, false)`

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,113 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+    auto req_columns = columns_context.requiredColumns();
+    for (const auto & req_column : req_columns)
+    {
+        std::sort(names.begin(), names.end());
```
Very strange code, you sort `names` on each iteration just to call `binary_search`?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,113 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+
+    builder.addSimpleTransform([&](const Block & stream_header)
+    {
+        return std::make_shared<FilterTransform>(stream_header, filter_actions, filter_column_name, true /* remove fake column */);
```
What is that fake column and why do we need to remove it?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,113 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+    std::set_difference(outer_set.begin(), outer_set.end(),
+        inner_set.begin(), inner_set.end(), std::inserter(result, result.begin()));
+
+    if (result.size() != 1)
```
Why should `result.size()` be always `1`?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -666,6 +721,17 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+
+            if (row_policy_data.needCare())
+            {
+                if (auto * source_step_with_filter = dynamic_cast<SourceStepWithFilter*>((plan.getRootNode()->step.get())))
```
What if that `step` is not derived from `SourceStepWithFilter`?

vitlibar replied:
Please add a comment here that it's an optimization if the storage can use filtering while reading (e.g. MergeTree).

vitlibar replied:
The conditions can be combined to be shorter:

```
if (plan.isInitialized() and row_policy_data)
{
    if (auto * source_step_with_filter = dynamic_cast<SourceStepWithFilter*>((plan.getRootNode()->step.get())))
        row_policy_data->addStorageFilter(source_step_with_filter);
}
```

ilejn replied:
It is not obvious that "The conditions can be combined to be shorter".
Could you explain why you are sure that going further with non initialized _plan_ is safe and makes sense?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -965,8 +1132,14 @@ void ReadFromMerge::convertingSourceStream(
+    QueryProcessingStage::Enum processed_stage,
+    RowPolicyData & row_policy_data)
 {
+    if (row_policy_data.needCare())
```
`needCare()` is not a very clear name. `row_policy_data.hasRowPolicy()` seems better.

ilejn replied:
Renamed, although it is usually considered bad practice if a method name contain class name (something like Vector::vector_size).
Nevertheless, let it be 'hasRowPolicy'.

vitlibar replied:
Better remove the method at all, see https://github.com/ClickHouse/ClickHouse/pull/50209#discussion_r1358108401

vitlibar commented on tests/queries/0_stateless/02763_row_policy_storage_merge.sql.j2:
```diff
@@ -0,0 +1,137 @@
+DROP ROW POLICY IF EXISTS 02763_filter_4 ON 02763_merge_merge_1;
+DROP ROW POLICY IF EXISTS 02763_filter_5 ON 02763_merge_fancycols;
+DROP ROW POLICY IF EXISTS 02763_filter_6 ON 02763_merge_fancycols;
+
```
Could you please write in a description to this PR a very simple example how it worked before your PR and how it's going to work from now on?

vitlibar replied:
I have made it by myself:

```
CREATE TABLE tbl1 (x UInt32) ENGINE=MergeTree ORDER BY tuple();
INSERT INTO tbl1 SELECT number AS x FROM numbers(10);
CREATE ROW POLICY r1 ON tbl1 USING x%2==0 TO ALL;
SELECT * FROM tbl1; -- applies the row policy
SELECT * FROM merge('default', 'tbl1'); -- doesn't apply the row policy
```

ilejn replied:
Original behavior is described in the issue
https://github.com/ClickHouse/ClickHouse/issues/50163
which is referred.

vitlibar replied:
My example is shorter :)

ilejn committed:
merge_row_policy: per code review

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -683,10 +749,10 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
-        }
 
-        if (!plan.isInitialized())
-            return {};
```
That check for `!plan.isInitialized()` on the outer level looked better.

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -965,8 +1138,14 @@ void ReadFromMerge::convertingSourceStream(
+    if (row_policy_data.hasRowPolicy())
+    {
+        row_policy_data.addFilterTransform(builder);
+    }
```
`convertingSourceStream()` is a function for converting the types of columns. It seems to me it would be better to call `row_policy_data.addFilterTransform()` directly from the caller function, i.e. from `ReadFromMerge::createSources()`

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+
+        actions_dag = expression_analyzer.getActionsDAG(false /* add_aliases */, false /* project_result */);
+        filter_actions = std::make_shared<ExpressionActions>(actions_dag,
+            ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
```
It seems you can extract everything you need from `filter_actions`, there is no need in `namesDifference()` function or in `RequiredSourceColumnsVisitor`:

```
        filter_input_columns = filter_actions->getRequiredColumns();

        const auto & sample_block = filter_actions->getSampleBlock();
        const auto & output_positions = filter_actions->getResultPositions();

        size_t num_result_columns = 0;

        for (size_t output_position : output_positions)
        {
            const auto & column_name = sample_block.getByPosition(output_position).name;
            if (std::find(filter_input_columns.begin(), filter_input_columns.end(), column_name) == filter_input_columns.end())
            {
                filter_column_name = column_name;
                ++num_result_columns;
            }
        }

        if (num_result_columns != 1)
            throw ...
```

ilejn replied:
Thanks, as I recall it was the initial approach.
I recheck if it is ok.

ilejn replied:
Hello @vitlibar , 
this code can substitute namesDifference, yes. Actually namesDifference implements exactly the same, but in slightly more elegant way.
Let me know if you have different opinion and I should switch to your code.

What I cannot understand is how this code can substitute RequiredSourceColumnsVisitor. Could you elaborate please? 
extendNames adds columns (to retrieve from storage) that are needed exclusively to evaluate row filter.

ilejn replied:
Ok, I see, RequiredSourceColumnsVisitor is not needed indeed, filter_actions->getRequiredColumns() is enough in for extendNames

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+{
+    assert(row_policy_filter_ptr);
+    ASTPtr expr = row_policy_filter_ptr->expression;
+
```
It's not necessary to use binary search here. The number of columns is not that big anyway.
Just store the input columns in the `init()` function and then use an ordinary linear search:

```
    for (const auto & filter_input_column : filter_input_columns)
    {
        if (std::find(names.begin(), names.end(), filter_input_column) == names.end())
            names.emplace_back(filter_input_column);
    }
```

vitlibar replied:
Your "binary search" doesn't seem to be faster than a linear search anyway. The complexity of a linear search is `O(N)`. You're using sort + binary search, the complexity is `O(N*log(N))`.

ilejn replied:
Assuming there is one column to filter?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -658,6 +700,19 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+                    table_name,
+                    RowPolicyFilterType::SELECT_FILTER),
+                storage,
+                modified_context);
```
If you move the code from `RowPolicyData::init()` to the constructor `RowPolicyData::RowPolicyData()` and then change the type of the `row_policy_data` variable to `std::optional<RowPolicyData>` then the code will look cleaner and you'll able to get rid of the function `RowPolicyData::hasRowPolicy()`

ilejn replied:
I doubt that suggested approach leads to cleaner code, because the decision if hasRowPolicy or not will be moved to outer scope.
Nevertheless, I will do it.

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+}
+
+// Add columns that needed to evaluate row policies
+// SELECT x from t  if  t has row policy
```
Do you have a test for the case described in this comment?

ilejn replied:
Yes, quite some actually.
e.g.
```
SELECT x, lc from merge(currentDatabase(), '02763_merge_fancycols') ORDER BY x;
```

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+
+void ReadFromMerge::RowPolicyData::addFilterTransform(QueryPipelineBuilder & builder)
+{
+    assert(row_policy_filter_ptr);
```
`std::optional<RowPolicyData>` (see my [other comment](https://github.com/ClickHouse/ClickHouse/pull/50209#discussion_r1358108401)) would allow to get rid of such asserts.

vitlibar reviewed (changes requested):
The PR is good, but it would be nice to improve the code in a few places.

ilejn committed:
merge_row_policy: some changes per code review

ilejn committed:
merge_row_policy: a comment

ilejn committed:
merge_row_policy: get rid of RequiredSourceColumnsVisitor

ilejn committed:
merge_row_policy: cleanup

ilejn commented:
Hello @vitlibar , I think I've addressed your comments, except binary search.
I've switched to NamesAndTypesList::getDifference in RowPolicyData ctor , which probably made code more natural.

ilejn commented:
Hello @vitlibar , do you have ideas/suggestions what should be improved further?
Should I rebase onto fresh master?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -789,6 +846,83 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+    auto storage_columns = storage_metadata_snapshot->getColumns();
+    auto needed_columns = storage_columns.getAllPhysical();
+
+    auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
```
Is it ok to use here `storage_columns.getAllPhysical()` as `needed_columns`? I mean what if a storage has an alias column and a row policy filter involves that alias?

ilejn replied:
> Is it ok to use here `storage_columns.getAllPhysical()` as `needed_columns`? I mean what if a storage has an alias column and a row policy filter involves that alias? Can you please check that case?

Very good point, quite possible that it is a mistake. Will check.

ilejn replied:
Indeed we have an issue here.
But the problem is deeper, Engine Merge is not very good at aliases in general, they seem do not work via function https://fiddle.clickhouse.com/4648fd3d-3c1a-462e-919b-90dafcafc131

ilejn replied:
The problem is even deeper ;) , looks like Engine Merge does not work with tables with different structures. And 'does not work' means gives wrong result.
https://fiddle.clickhouse.com/27a3d1d1-51a4-491c-b6ad-5421c11d1f0b

ilejn replied:
Summing up issue with alias.
I've fixed it for cases where Engine Merge is used directly, not via table function merge(). Works ok. Not pushed yet.
I haven't fixed it for merge() yet.
How it looks https://pastila.nl/?00d38392/d68be2ab03fb64572f5a3f1b3bd383b2#2RTThmq3iM6cf1C1nLQVUg==

Alternatively we can 

1. expose alias via merge table function - trivial fix. Can be treated as improvement, but breaks current behavior.
2. go with this PR further as is and fix alias based row policy for merge() in a separated PR.

ilejn replied:
Considering we have  _asterisk_include_alias_columns_ the first option seems fine.
Currently table function merge() does not expose aliases.
@vitlibar , what do you think?

vitlibar replied:
> expose alias via merge table function - trivial fix. Can be treated as improvement, but breaks current behavior.

I suppose ideally the `merge()` function should return aliases only if that alias is requested explicitly; or if asterisk is used and ` asterisk_include_alias_columns` is set. Because it's better to be consistent with how we treat aliases for normal tables.

vitlibar replied:
Of course a better solution is always better, I just suggested to split work between two PRs. But ok, no problem, do it the way you like.

ilejn replied:
The last push brought many changes in term of lines of code, but actually changes are not significant.
I've introduced processAliases method because I have to add new names before taking care of aliases and it is too much clutter for initializePipeline().
Besides this, I renamed convertingSourceStream to convertAndFilterSourceStream , because filtering is needed in-between.

ilejn replied:
asterisk_include_alias_columns is left out of the scope despite my initial intention (and claim).

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -789,6 +846,83 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+
+void ReadFromMerge::RowPolicyData::addStorageFilter(SourceStepWithFilter * step)
+{
+    assert(row_policy_filter_ptr);
```
Not really necessary because you've already checked that in the constructor

ilejn replied:
> Not really necessary because you've already checked that in the constructor

That's an idea of assert.
We state that we rely on something that is provided by another code.

vitlibar replied:
We usually don't assert pointers to be not null, because if they happen to be null the first access by that pointer will stop the program immediately with a fatal error - and it's almost the same as if the program fails on `assert()`. So it doesn't make much sense in checking pointers with `assert()`.

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -789,6 +846,83 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+    if (!deleted.empty() || added.size() != 1)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Cannot determine row level filter");
```
Please add information about the size of `deleted` and `added` here to the error message to help diagnostics if something will go wrong.

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -789,6 +846,83 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+
+void ReadFromMerge::RowPolicyData::extendNames(Names & names)
+{
+    std::sort(names.begin(), names.end());
```
I find it quite surprising that the function `extendNames()` reorders a list of names passed to that function as an argument.

vitlibar replied:
You could use `std::string_view` to optimize it a little bit and also to preserve the original order of `names`:

```
boost::container::flat_set<std::string_view> names_set(names.begin(), names.end());
NameSet added_names;

for (const auto & req_column : filter_actions->getRequiredColumns())
{
    if (!names_set.contains(req_column))
         added_names.emplace(req_column);
}
```

ilejn committed:
merge_row_policy: aliases slightly work

ilejn committed:
merge_row_policy: processAliases introduced

ilejn committed:
merge_row_policy: cleanup, traces removed

ilejn committed:
merge_row_policy: style fix, cleanup

vitlibar commented on src/Storages/StorageMerge.h:
```diff
@@ -177,31 +177,39 @@ class ReadFromMerge final : public SourceStepWithFilter
+        const StorageWithLockAndName & storage_with_lock,
+        Aliases & aliases,
+        const Block & sample_block,
+        ContextMutablePtr modified_context);
```
What exactly the function `processAliases()` does? I know it's named `processAliases` but how it processes them? A comment here would be useful.

ilejn replied:
I've moved this part of code as is, but Ok, I'll try to add some comments.

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -616,22 +592,109 @@ SelectQueryInfo ReadFromMerge::getModifiedQueryInfo(const SelectQueryInfo & quer
+    }
+
+    Aliases aliases;
+    processAliases(real_column_names, storage_with_lock, aliases, sample_block, modified_context);
```
Maybe it's better to rename `processAliases` to `getAliases()` and make it return `Aliases`?

vitlibar commented on src/Storages/StorageMerge.cpp:
```diff
@@ -616,22 +592,109 @@ SelectQueryInfo ReadFromMerge::getModifiedQueryInfo(const SelectQueryInfo & quer
+    }
+    if (!column_names_as_aliases.empty())
+    {
+        real_column_names = column_names_as_aliases;
```
So it replaces `real_column_names` only if `!column_names_as_aliases.empty()`
Such a complicated logic.

vitlibar replied:
Let's make the function `processAliases()` return `column_names_as_aliases` as is, and do replace `read_column_names` not here but in the function which calls `processAliases()`

vitlibar replied:
You've made an interference already, it doesn't add too much interference if we pass arguments this way or another. I'm ok with passing results via reference parameters. For me it's just better to avoid parameters which are input and output at the same time - just to make the life easier. So why can't we make `column_names_as_aliases` a pure output parameter and do this logic in `createSource()`:

```
... createSource(...)
{
    ...
    Aliases aliases;
    Names column_names_as_aliases;
    processAliases(context, storage_with_lock, sample_block, modified_context, aliases, column_names_as_aliases);
    if (!column_names_as_aliases.empty())
        real_column_names = column_names_as_aliases;
    ...
```

ilejn replied:
May be because this logic is a part of processing aliases ? ;)

ilejn committed:
merge_row_policy: some comments

vitlibar commented:
Please resolve the conflict

ilejn commented:
> Please resolve the conflict

The changes are rather significant.

ilejn committed:
Merge remote-tracking branch 'origin/master' into merge_row_policy

merge_row_policy: changes because of refactoring in master

ilejn committed:
merge_row_policy: cleanup after merge

ilejn committed:
merge_row_policy: alias test separated out and disabled if analyzer

vitlibar commented on tests/analyzer_tech_debt.txt:
```diff
@@ -42,6 +42,7 @@
 02003_WithMergeableStateAfterAggregationAndLimit_LIMIT_BY_LIMIT_OFFSET
 02404_memory_bound_merging
 02725_agg_projection_resprect_PK
+02763_row_policy_storage_merge_alias
```
Why?

ilejn replied:
Basically because of 
```
$ grep merge_alias analyzer_tech_debt.txt 
01214_test_storage_merge_aliases_with_where
01925_test_storage_merge_aliases
```
Aliases in Merge do not work if analyzer enabled.

ilejn commented:
> The build is broken ([see](https://s3.amazonaws.com/clickhouse-builds/PRs/50209/f226fa685bc55f7b81f4838a45a6037a4202dc51/binary_tidy/build_log.log)). Can you please fix it?

Yes, sure.

ilejn committed:
merge_row_policy: make clang-tidy happy

ilejn commented:
@vitlibar 
I've got rid of processAliases because in refactored version there fewer reasons for it.
And no problem of non-const references as a bonus.
To keep balance ;)  I've introduced rvalue reference for real_column_names which might be strange but seems the best option to me.

ilejn commented:
Test failures are not caused by the proposed changes.
